src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -1346,4 +1368,149 @@ public static MethodHandle makeConcatWithTemplateGetters(
+                        @Override
+                        public void accept(ClassBuilder clb) {
+                            clb.withFlags(AccessFlag.PUBLIC, AccessFlag.FINAL, AccessFlag.SUPER, AccessFlag.SYNTHETIC)
+                                .withMethodBody(METHOD_NAME,
+                                        MethodTypeDesc.ofDescriptor(args.toMethodDescriptorString()),
+                                        ClassFile.ACC_FINAL | ClassFile.ACC_PUBLIC | ClassFile.ACC_STATIC,
+                                        generateMethod(constants, args));
+                    }});
+
+            try {
+                Class<?> innerClass = lookup.defineHiddenClass(classBytes, true, STRONG).lookupClass();
+                DUMPER.dumpClass(className, innerClass, classBytes);
+                MethodHandle mh = lookup.findStatic(innerClass, METHOD_NAME, args);
+                return mh;
+            } catch (Exception e) {
+                DUMPER.dumpFailedClass(className, classBytes);
+                throw new StringConcatException("Exception while spinning the class", e);
+            }
+        }
+
+        private static Consumer<CodeBuilder> generateMethod(String[] constants, MethodType args) {
+            return new Consumer<CodeBuilder>() {
+                @Override
+                public void accept(CodeBuilder cb) {
+                    cb.new_(STRING_BUILDER);
+                    cb.dup();
+                    cb.invokespecial(STRING_BUILDER, "<init>", MethodTypeDesc.ofDescriptor("()V"));
+
+                    // At this point, we have a blank StringBuilder on stack, fill it in with .append calls.
+                    {
+                        int off = 0;
+                        for (int c = 0; c < args.parameterCount(); c++) {
+                            if (constants[c] != null) {
+                                cb.ldc(constants[c]);
+                                cb.invokevirtual(STRING_BUILDER, "append", APPEND_STRING_TYPE);
+                            }
+                            Class<?> cl = args.parameterType(c);
+                            off += load(cb, cl, off);
+                            MethodTypeDesc desc = getSBAppendDesc(cl);
+                            cb.invokevirtual(STRING_BUILDER, "append", desc);
+                        }
+                        if (constants[constants.length - 1] != null) {
+                            cb.ldc(constants[constants.length - 1]);
+                            cb.invokevirtual(STRING_BUILDER, "append", APPEND_STRING_TYPE);
+                        }
+                    }
+
+                    cb.invokevirtual(STRING_BUILDER, "toString", MethodTypeDesc.ofDescriptor("()Ljava/lang/String;"));
+                    cb.areturn();
+                }
+            };
+        }
+
+        // Load the argument of type cl at slot onto stack, return the number of argument stack slots consumed.
+        private static int load(CodeBuilder cb, Class<?> cl, int slot) {
```

Comment:
You can replace all calls to `load(cb, cl, slot)` with `cb.loadInstruction(TypeKind.from(cl), slot)`, and the retrn slot count can be accessed by `TypeKind::slotSize`

Reply:
Thanks, I've tried to simplify.